packages/fxa-auth-server/scripts/stripe-products-and-plans-to-firestore-documents/plan-language-tags-guesser.ts

```diff
@@ -0,0 +1,160 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import { v2 as CloudTranslate } from '@google-cloud/translate';
+import superagent from 'superagent';
+import { Stripe } from 'stripe';
+
+/**
+ * This lib attempts to detect the (human) language used in a Stripe plan's
+ * metadata.
+ *
+ * The language will be determined with one or more of the following:
+ *  - the language detected by Google Cloud Translation from the plan's product details metadata
+ *  - the plan's title
+ *  - the plan's currency
+ */
+
+const MIN_CONFIDENCE = 0.51;
+export const PLAN_EN_LANG_ERROR = 'Plan specific en metadata';
+const { Translate } = CloudTranslate;
+const translate = new Translate();
+let locales: string[];
+
+const initLocales = async () => {
+  if (!locales) {
+    const localesRes = await superagent.get(
+      'https://raw.githubusercontent.com/unicode-org/cldr-json/main/cldr-json/cldr-core/availableLocales.json'
+    );
+    locales = JSON.parse(localesRes.text).availableLocales.modern.map(
+      (x: string) => x.toLowerCase()
+    );
+  }
+};
+
+const getMetadataProductDetails = (metadata: Stripe.Metadata) =>
+  Object.entries(metadata)
+    .reduce((acc, [k, v]) => {
+      if (k.startsWith('product:detail')) {
+        acc.push(v);
+      }
+      return acc;
+    }, [])
+    .join(' ');
+
+const searchTagsInLocales = (tags: string[]) => {
+  for (const t of tags) {
+    if (locales.includes(t.toLowerCase())) {
+      return t;
+    }
+  }
+};
+
+const findLocaleInTitle = (lang: string, planTitle: string) => {
+  if (!planTitle) {
+    return lang;
+  }
+
+  const words = planTitle.split(' ');
+
+  // the entire tag is in the title
+  const potentialLanguageTags = words.filter((x) =>
+    x.toLowerCase().startsWith(`${lang}-`)
+  );
+  const tagFromTitle = searchTagsInLocales(potentialLanguageTags);
+  if (tagFromTitle) {
+    return tagFromTitle;
+  }
+
+  // a sub tag is in the title
+  const searchTerms = words
+    .filter((x) => x.toLowerCase() !== lang)
+    .map((t) => `${lang}-${t}`);
+  const tagWithSubtagFromTitle = searchTagsInLocales(searchTerms);
+  if (tagWithSubtagFromTitle) {
+    return tagWithSubtagFromTitle;
+  }
+
+  // default to the detected lang
+  return lang;
+};
+
+const mapCurrencyToLocale = (currency: string) => {
+  // there is no direct connection between a currency and a localised
+  // language, but with the special case of Switzerland (the only one at the
+  // time of writing), this is the sure way to detect for Swiss language
+  // tags.
+  const currencyToLocaleMap: { [key: string]: string } = { chf: 'CH' };
+  if (currencyToLocaleMap[currency.toLowerCase()]) {
+    return currencyToLocaleMap[currency.toLowerCase()];
+  }
+};
+
+const formatLanguageTag = (tag: string) =>
+  tag
+    .split('-')
+    .map((x, idx) => (idx === 0 ? x.toLowerCase() : x))
+    .join('-');
```

Comment:
It's making the language all lowercase as the `map` is over the result of the `split`.

Reply:
Ah okay thanks.

Why do you only want to put the language subtag to lower case and not both the language and extended language subtags?

Looking at `searchTagsInLocales`, you convert the entire tag to lowercase before looking in `locales`, whose tags have also been converted to lowercase.

Reply:
It's just the convention I've noticed in the CLDR locales.  I'm not sure if it makes any difference, e.g. other language tag handling libraries might perform a to upper or lower case before comparisons.

For comparisons in this patch, since we don't enforce any sort of format on the plan nickname, I lowercase the values before a comparison of them.